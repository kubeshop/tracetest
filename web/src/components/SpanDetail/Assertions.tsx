import {useState} from 'react';
import {Typography} from 'antd';
import {PlusOutlined} from '@ant-design/icons';
import {ISpan} from 'types/Span.types';
import SkeletonTable from 'components/SkeletonTable';
import AssertionsResultTable from '../AssertionsTable';
import * as S from './SpanDetail.styled';
import CreateAssertionModal from '../CreateAssertionModal';
import {IAssertionResultList} from '../../types/Assertion.types';

interface IProps {
  span?: ISpan;
  testId?: string;
  resultId?: string;
  assertionsResultList: IAssertionResultList[];
}

const Assertion: React.FC<IProps> = ({assertionsResultList, testId, span, resultId}) => {
  const [isModalOpen, setIsModalOpen] = useState(false);

  return (
    <SkeletonTable loading={!span}>
      <S.AssertionActionsContainer>
        <S.AddAssertionButton
          data-cy="add-assertion-button"
          icon={<PlusOutlined />}
          onClick={() => setIsModalOpen(true)}
        >
          Add Assertion
        </S.AddAssertionButton>
      </S.AssertionActionsContainer>
      {assertionsResultList.length ? (
        assertionsResultList
          .sort((a, b) => (a.assertion.assertionId > b.assertion.assertionId ? -1 : 1))
          .map(({assertion, assertionResultList}, index) => (
            <AssertionsResultTable
              key={assertion.assertionId}
              assertionResults={assertionResultList}
              sort={index + 1}
              assertion={assertion}
              span={span!}
              resultId={resultId!}
              testId={testId!}
            />
          ))
      ) : (
        <S.DetailsEmptyStateContainer data-cy="empty-assertion-table">
          <S.DetailsTableEmptyStateIcon />
          <Typography.Text disabled>No Data</Typography.Text>
        </S.DetailsEmptyStateContainer>
      )}

      <CreateAssertionModal
        open={isModalOpen}
        onClose={() => setIsModalOpen(false)}
        span={span!}
        testId={testId!}
        resultId={resultId!}
      />
    </SkeletonTable>
  );
};

export default Assertion;
