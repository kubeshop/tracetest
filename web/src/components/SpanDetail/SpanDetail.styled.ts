import {Divider, Typography} from 'antd';
import styled from 'styled-components';

export const Header = styled.div`
  padding: 16px 12px 0;
  min-width: 270px;
`;

export const HeaderCheck = styled(Typography.Text)`
  color: ${({theme}) => theme.color.text};
  font-size: ${({theme}) => theme.size.sm};
  margin-right: 8px;
`;

export const HeaderDot = styled.span<{$passed: boolean}>`
  background-color: ${({$passed, theme}) => ($passed ? theme.color.success : theme.color.error)};
  height: 10px;
  width: 10px;
  display: inline-block;
  margin-right: 4px;
  line-height: 0;
  vertical-align: -0.1em;
  border-radius: 50%;
`;

export const HeaderDivider = styled(Divider)`
  margin: 16px 0;
`;

export const HeaderItem = styled.div`
  align-items: center;
  color: ${({theme}) => theme.color.text};
  display: flex;
  font-size: ${({theme}) => theme.size.md};
  margin-right: 8px;
`;

export const HeaderItemText = styled(Typography.Text)`
  color: inherit;
  margin-left: 5px;
`;

export const HeaderTitle = styled(Typography.Title)`
  && {
    margin: 0;
  }
`;

export const Column = styled.div`
  align-items: flex-start;
  display: flex;
  flex-direction: column;
  margin-bottom: 4px;
`;

export const Row = styled.div`
  align-items: center;
  display: flex;
  margin-bottom: 4px;
`;

export const AttributesContainer = styled.div<{$top: number}>`
  height: calc(100% - ${({$top}) => `${$top}px`});
  overflow-y: scroll;
`;

export const SearchContainer = styled.div`
  padding: 0 12px;
`;
