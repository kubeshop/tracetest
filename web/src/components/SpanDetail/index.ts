import TestAttributeRow from './TestAttributeRow/TestAttributeRow';
import TestSubHeader from './TestSubHeader/TestSubHeader';
import TraceAttributeRow from './TraceAttributeRow/TraceAttributeRow';
import TraceSubHeader from './TraceSubHeader/TraceSubHeader';

// eslint-disable-next-line no-restricted-exports
export {default} from './SpanDetail';
export {TestAttributeRow, TestSubHeader, TraceAttributeRow, TraceSubHeader};
