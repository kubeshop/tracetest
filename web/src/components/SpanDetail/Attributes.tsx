import {useEffect, useRef, useState} from 'react';
import AttributeList from 'components/AttributeList';
import {OtelReference} from 'components/TestSpecForm/hooks/useGetOTELSemanticConventionAttributesInfo';
import TestRunOutput from 'models/TestRunOutput.model';
import {TSpanFlatAttribute} from 'types/Span.types';
import {TAnalyzerError, TTestSpecSummary} from 'types/TestRun.types';
import {IPropsAttributeRow} from './SpanDetail';
import * as S from './SpanDetail.styled';

interface IProps {
  attributeList: TSpanFlatAttribute[];
  searchText?: string;
  semanticConventions: OtelReference;
  analyzerErrors?: TAnalyzerError[];
  testSpecs?: TTestSpecSummary;
  testOutputs?: TestRunOutput[];
  onCreateTestSpec(attribute: TSpanFlatAttribute): void;
  onCreateOutput(attribute: TSpanFlatAttribute): void;
  AttributeRowComponent: React.ComponentType<IPropsAttributeRow>;
}

const Attributes = ({
  attributeList,
  searchText,
  semanticConventions,
  analyzerErrors,
  testSpecs,
  testOutputs,
  onCreateTestSpec,
  onCreateOutput,
  AttributeRowComponent,
}: IProps) => {
  const containerRef = useRef<HTMLDivElement>(null);
  const [topPosition, setTopPosition] = useState(0);

  useEffect(() => {
    setTopPosition(containerRef?.current?.offsetTop ?? 0);
  }, [attributeList]);

  return (
    <S.AttributesContainer $top={topPosition} ref={containerRef}>
      <AttributeList
        attributeList={attributeList}
        searchText={searchText}
        semanticConventions={semanticConventions}
        analyzerErrors={analyzerErrors}
        testSpecs={testSpecs}
        testOutputs={testOutputs}
        onCreateTestSpec={onCreateTestSpec}
        onCreateOutput={onCreateOutput}
        AttributeRowComponent={AttributeRowComponent}
      />
    </S.AttributesContainer>
  );
};

export default Attributes;
