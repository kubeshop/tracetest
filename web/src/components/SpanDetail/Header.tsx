import {SettingOutlined, ToolOutlined} from '@ant-design/icons';
import {Space} from 'antd';
import * as SSpanNode from 'components/Visualization/components/DAG/BaseSpanNode/BaseSpanNode.styled';
import {SemanticGroupNamesToText} from 'constants/SemanticGroupNames.constants';
import Span from 'models/Span.model';
import SpanService from 'services/Span.service';
import * as S from './SpanDetail.styled';
import {StepsID} from '../GuidedTour/testRunSteps';

interface IProps {
  span?: Span;
}

const Header = ({span}: IProps) => {
  const {kind, name, service, system, type} = SpanService.getSpanInfo(span);

  if (!span) {
    return (
      <S.Header data-tour={StepsID.SpanDetails}>
        <S.HeaderTitle level={3}>Span Attributes</S.HeaderTitle>
      </S.Header>
    );
  }

  return (
    <S.Header data-tour={StepsID.SpanDetails}>
      <S.Column>
        <Space>
          <SSpanNode.BadgeType $hasMargin count={SemanticGroupNamesToText[type]} $type={type} />
        </Space>
        <S.HeaderTitle level={3}>{name}</S.HeaderTitle>
      </S.Column>
      <S.Column>
        <S.HeaderItem>
          <SettingOutlined />
          <S.HeaderItemText>{`${service} ${kind}`}</S.HeaderItemText>
        </S.HeaderItem>
        {Boolean(system) && (
          <S.HeaderItem>
            <ToolOutlined />
            <S.HeaderItemText>{system}</S.HeaderItemText>
          </S.HeaderItem>
        )}
      </S.Column>
    </S.Header>
  );
};

export default Header;
