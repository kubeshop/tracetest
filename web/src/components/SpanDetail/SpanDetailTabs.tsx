import {Tabs} from 'antd';
import {capitalize} from 'lodash';
import React, {useMemo} from 'react';

import AttributeList from 'components/AttributeList';
import TraceAnalyticsService from 'services/Analytics/TraceAnalytics.service';
import SpanAttributeService from 'services/SpanAttribute.service';
import {ISpanDetailsComponentProps} from './SpanDetail';

const SpanDetailTabs: React.FC<ISpanDetailsComponentProps> = ({
  span: {attributeList = [], type} = {},
  onCreateAssertion,
  assertions,
}) => {
  const sectionList = useMemo(
    () => SpanAttributeService.getSpanAttributeSectionsList(attributeList, type!),
    [attributeList, type]
  );

  return (
    <Tabs data-cy="span-details-attributes" onChange={tabName => TraceAnalyticsService.onChangeTab(tabName)}>
      {sectionList.map(({section, attributeList: attrList}) => (
        <Tabs.TabPane tab={capitalize(section)} key={section}>
          <AttributeList assertions={assertions} attributeList={attrList} onCreateAssertion={onCreateAssertion} />
        </Tabs.TabPane>
      ))}
    </Tabs>
  );
};

export default SpanDetailTabs;
