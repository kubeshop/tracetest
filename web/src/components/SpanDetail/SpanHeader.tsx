import {Typography} from 'antd';
import GuidedTourService, {GuidedTours} from '../../services/GuidedTour.service';
import {Steps} from '../GuidedTour/traceStepList';
import * as S from './SpanDetail.styled';

interface IProps {
  title: string;
}

const SpanHeader: React.FC<IProps> = ({title}) => (
  <S.SpanHeader data-tour={GuidedTourService.getStep(GuidedTours.Trace, Steps.Details)}>
    <S.SpanHeaderTitle>Span Details</S.SpanHeaderTitle>
    <Typography.Text type="secondary">{title}</Typography.Text>
  </S.SpanHeader>
);

export default SpanHeader;
