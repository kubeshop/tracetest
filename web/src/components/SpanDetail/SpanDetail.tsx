import {noop} from 'lodash';
import {useCallback, useMemo, useState} from 'react';

import SearchInput from 'components/SearchInput';
import {
  OtelReference,
  useGetOTELSemanticConventionAttributesInfo,
} from 'components/TestSpecForm/hooks/useGetOTELSemanticConventionAttributesInfo';
import {useTestSpecForm} from 'components/TestSpecForm/TestSpecForm.provider';
import {CompareOperatorSymbolMap} from 'constants/Operator.constants';
import useSpanData from 'hooks/useSpanData';
import Span from 'models/Span.model';
import TestOutput from 'models/TestOutput.model';
import TestRunOutput from 'models/TestRunOutput.model';
import {useTestOutput} from 'providers/TestOutput/TestOutput.provider';
import TraceAnalyticsService from 'services/Analytics/TestRunAnalytics.service';
import AssertionService from 'services/Assertion.service';
import SpanService from 'services/Span.service';
import SpanAttributeService from 'services/SpanAttribute.service';
import {TSpanFlatAttribute} from 'types/Span.types';
import {TAnalyzerError, TTestSpecSummary} from 'types/TestRun.types';
import Attributes from './Attributes';
import Header from './Header';
import * as S from './SpanDetail.styled';

export interface IPropsAttributeRow {
  attribute: TSpanFlatAttribute;
  searchText?: string;
  semanticConventions: OtelReference;
  analyzerErrors?: TAnalyzerError[];
  testSpecs?: TTestSpecSummary;
  testOutputs?: TestRunOutput[];
  onCreateTestSpec(attribute: TSpanFlatAttribute): void;
  onCreateOutput(attribute: TSpanFlatAttribute): void;
}

export interface IPropsSubHeader {
  analyzerErrors?: TAnalyzerError[];
  testSpecs?: TTestSpecSummary;
  testOutputs?: TestRunOutput[];
}

interface IProps {
  onCreateTestSpec?(): void;
  searchText?: string;
  span?: Span;
  AttributeRowComponent: React.ComponentType<IPropsAttributeRow>;
  SubHeaderComponent: React.ComponentType<IPropsSubHeader>;
}

const SpanDetail = ({onCreateTestSpec = noop, searchText, span, AttributeRowComponent, SubHeaderComponent}: IProps) => {
  const {analyzerErrors, testSpecs, testOutputs} = useSpanData(span?.id ?? '');
  const {open} = useTestSpecForm();
  const {onNavigateAndOpen} = useTestOutput();
  const [search, setSearch] = useState('');
  const semanticConventions = useGetOTELSemanticConventionAttributesInfo();

  const handleCreateTestSpec = useCallback(
    ({value, key}: TSpanFlatAttribute) => {
      TraceAnalyticsService.onAddAssertionButtonClick();
      const selector = SpanService.getSelectorInformation(span!);

      open({
        isEditing: false,
        selector,
        defaultValues: {
          assertions: [
            {
              left: `attr:${key}`,
              comparator: CompareOperatorSymbolMap.EQUALS,
              right: AssertionService.extractExpectedString(value) || '',
            },
          ],
          selector,
        },
      });

      onCreateTestSpec();
    },
    [onCreateTestSpec, open, span]
  );

  const handleCreateOutput = useCallback(
    ({key}: TSpanFlatAttribute) => {
      TraceAnalyticsService.onAddAssertionButtonClick();
      const selector = SpanService.getSelectorInformation(span!);

      const output = TestOutput({
        selector,
        selectorParsed: {query: selector},
        name: key,
        value: `attr:${key}`,
      });

      onNavigateAndOpen({...output, spanId: span!.id});
    },
    [onNavigateAndOpen, span]
  );

  const handleOnSearch = useCallback((value: string) => {
    setSearch(value);
  }, []);

  const filteredAttributes = useMemo(
    () => SpanAttributeService.filterAttributes(span?.attributeList ?? [], search, semanticConventions),
    [span?.attributeList, search, semanticConventions]
  );

  return (
    <>
      <Header span={span} />
      <SubHeaderComponent analyzerErrors={analyzerErrors} testSpecs={testSpecs} testOutputs={testOutputs} />
      <S.HeaderDivider />

      <S.SearchContainer data-cy="attributes-search-container">
        <SearchInput placeholder="Search attributes" onSearch={handleOnSearch} width="100%" />
      </S.SearchContainer>

      <Attributes
        attributeList={filteredAttributes}
        searchText={searchText}
        semanticConventions={semanticConventions}
        analyzerErrors={analyzerErrors}
        testSpecs={testSpecs}
        testOutputs={testOutputs}
        onCreateTestSpec={handleCreateTestSpec}
        onCreateOutput={handleCreateOutput}
        AttributeRowComponent={AttributeRowComponent}
      />
    </>
  );
};

export default SpanDetail;
