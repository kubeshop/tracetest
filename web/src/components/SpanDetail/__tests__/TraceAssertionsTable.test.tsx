import {render} from '@testing-library/react';
import {ReduxWrapperProvider} from '../../../redux/ReduxWrapperProvider';
import {TestingModels} from '../../../utils/TestingModels';
import SpanDetail from '../SpanDetail';

test('Layout', () => {
  const result = render(
    <ReduxWrapperProvider>
      <SpanDetail resultId={TestingModels.resultId} testId={TestingModels.testId} span={TestingModels.span} />
    </ReduxWrapperProvider>
  );
  expect(result.container).toMatchSnapshot();
});
