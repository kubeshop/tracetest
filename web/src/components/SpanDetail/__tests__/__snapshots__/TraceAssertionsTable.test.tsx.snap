// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout 1`] = `
<div>
  <div
    class="sc-ieecCq FiStd"
  >
    <div
      class="sc-bqiRlB iwrCuh"
    >
      <h5
        class="ant-typography sc-ksdxgE ipQGyD"
      >
        Span Details
      </h5>
      <span
        class="ant-typography ant-typography-secondary"
      >
        HTTP •  • 
      </span>
    </div>
    <div
      class="ant-tabs ant-tabs-top sc-hBUSln kUdfoR"
      data-cy="span-details-attributes"
      style="margin-bottom: 150px; overflow: hidden;"
    >
      <div
        class="ant-tabs-nav"
        role="tablist"
      >
        <div
          class="ant-tabs-nav-wrap"
        >
          <div
            class="ant-tabs-nav-list"
            style="transform: translate(0px, 0px);"
          >
            <div
              class="ant-tabs-tab ant-tabs-tab-active"
            >
              <div
                aria-controls="rc-tabs-test-panel-span-assertion"
                aria-selected="true"
                class="ant-tabs-tab-btn"
                id="rc-tabs-test-tab-span-assertion"
                role="tab"
                tabindex="0"
              >
                Assertion
              </div>
            </div>
            <div
              class="ant-tabs-tab"
            >
              <div
                aria-controls="rc-tabs-test-panel-span-request"
                aria-selected="false"
                class="ant-tabs-tab-btn"
                id="rc-tabs-test-tab-span-request"
                role="tab"
                tabindex="0"
              >
                Request
              </div>
            </div>
            <div
              class="ant-tabs-tab"
            >
              <div
                aria-controls="rc-tabs-test-panel-span-response"
                aria-selected="false"
                class="ant-tabs-tab-btn"
                id="rc-tabs-test-tab-span-response"
                role="tab"
                tabindex="0"
              >
                Response
              </div>
            </div>
            <div
              class="ant-tabs-tab"
            >
              <div
                aria-controls="rc-tabs-test-panel-span-attribute-list"
                aria-selected="false"
                class="ant-tabs-tab-btn"
                id="rc-tabs-test-tab-span-attribute-list"
                role="tab"
                tabindex="0"
              >
                Attribute list
              </div>
            </div>
            <div
              class="ant-tabs-ink-bar ant-tabs-ink-bar-animated"
            />
          </div>
        </div>
        <div
          class="ant-tabs-nav-operations ant-tabs-nav-operations-hidden"
        >
          <button
            aria-controls="rc-tabs-test-more-popup"
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-hidden="true"
            class="ant-tabs-nav-more"
            id="rc-tabs-test-more"
            style="visibility: hidden; order: 1;"
            tabindex="-1"
            type="button"
          >
            <span
              aria-label="ellipsis"
              class="anticon anticon-ellipsis"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="ellipsis"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M176 511a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0z"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
      <div
        class="ant-tabs-content-holder"
      >
        <div
          class="ant-tabs-content ant-tabs-content-top"
        >
          <div
            aria-hidden="false"
            aria-labelledby="rc-tabs-test-tab-span-assertion"
            class="ant-tabs-tabpane ant-tabs-tabpane-active"
            id="rc-tabs-test-panel-span-assertion"
            role="tabpanel"
            tabindex="0"
          >
            <div>
              <div
                class="sc-dJjYzT jydkUB"
              >
                <button
                  class="ant-btn ant-btn-primary sc-hGPBjI ijyrZa"
                  data-cy="add-assertion-button"
                  type="button"
                >
                  <span
                    aria-label="plus"
                    class="anticon anticon-plus"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="plus"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <defs>
                        <style />
                      </defs>
                      <path
                        d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z"
                      />
                      <path
                        d="M176 474h672q8 0 8 8v60q0 8-8 8H176q-8 0-8-8v-60q0-8 8-8z"
                      />
                    </svg>
                  </span>
                  <span>
                    Add Assertion
                  </span>
                </button>
              </div>
              <div
                class="sc-fFeiMQ dGJwRW"
                data-cy="empty-assertion-table"
              >
                <img
                  class="sc-bkkeKt"
                  src="SpanAssertionsEmptyState.svg"
                />
                <span
                  class="ant-typography ant-typography-disabled"
                >
                  Add assertion to see result here.
                </span>
              </div>
            </div>
          </div>
          <div
            aria-hidden="true"
            aria-labelledby="rc-tabs-test-tab-span-request"
            class="ant-tabs-tabpane"
            id="rc-tabs-test-panel-span-request"
            role="tabpanel"
            style="display: none;"
            tabindex="-1"
          />
          <div
            aria-hidden="true"
            aria-labelledby="rc-tabs-test-tab-span-response"
            class="ant-tabs-tabpane"
            id="rc-tabs-test-panel-span-response"
            role="tabpanel"
            style="display: none;"
            tabindex="-1"
          />
          <div
            aria-hidden="true"
            aria-labelledby="rc-tabs-test-tab-span-attribute-list"
            class="ant-tabs-tabpane"
            id="rc-tabs-test-panel-span-attribute-list"
            role="tabpanel"
            style="display: none;"
            tabindex="-1"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
